File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/2217-step-by-step-directions-from-a-binary-tree-node-to-another.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void parentTrack(TreeNode* root, unordered_map<TreeNode*, TreeNode*> &mp, int startValue, int destValue){
        if (!root) return ;
        if (root->left) mp[root->left] = root;
        if (root->right) mp[root->right] = root;
        parentTrack(root->left, mp, startValue, destValue);
        parentTrack(root->right, mp, startValue, destValue);
        return ;
    }
    TreeNode* getNode(TreeNode* root, int startValue) {
        if (!root) return nullptr;
        if (root->val == startValue) return root;
        TreeNode* lt = getNode(root->left, startValue);
        TreeNode* rt = getNode(root->right, startValue);
        if (lt) return lt;
        return rt;
    }
    bool solve(TreeNode* root, int &destValue, unordered_map<TreeNode*, TreeNode*> &mp, 
        string& path, unordered_set<int> &vis){
        if (!root) return false;
        vis.insert(root->val);
        if (root->val == destValue) return true;
        if (root->left && !vis.count(root->left->val)){
            path.push_back('L');
            if (solve(root->left, destValue, mp, path, vis)) return true;
        }
        if (root->right && !vis.count(root->right->val)){
            path.push_back('R');
            if (solve(root->right, destValue, mp, path, vis)) return true;
        }
        if (mp[root] && !vis.count(mp[root]->val)){
            path.push_back('U');
            if (solve(mp[root], destValue, mp, path, vis)) return true;
        }
        path.pop_back();
        return false;
    }
    string getDirections(TreeNode* root, int startValue, int destValue) {
        unordered_map<TreeNode*, TreeNode*> mp;
        unordered_set<int> vis;
        parentTrack(root, mp, startValue, destValue);
        TreeNode* node = getNode(root, startValue);
        string path;
        if (solve(node, destValue, mp, path, vis)) return path;
        return "";
    }
};

File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/2217-step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def parentTrack(self, root, mp):
        if not root:
            return
        if root.left:
            mp[root.left] = root
        if root.right:
            mp[root.right] = root
        self.parentTrack(root.left, mp)
        self.parentTrack(root.right, mp)
    
    def getNode(self, root, startValue):
        if not root:
            return None
        lt = self.getNode(root.left, startValue)
        rt = self.getNode(root.right, startValue)
        if root.val == startValue:
            return root
        if lt:
            return lt
        return rt

    def solve(self, root, destValue, mp, path, vis):
        if not root:
            return False
        vis[root.val] = True
        if root.val == destValue:
            return True
        if root.left and root.left.val not in vis:
            path.append('L')
            if self.solve(root.left, destValue, mp, path, vis):
                return True
        if root.right and root.right.val not in vis:
            path.append('R')
            if self.solve(root.right, destValue, mp, path, vis):
                return True
        if root in mp and mp[root].val not in vis:
            path.append('U')
            if self.solve(mp[root], destValue, mp, path, vis):
                return True
        path.pop()
        return False

    def getDirections(self, root, startValue, destValue):
        mp = {}
        vis = {}
        self.parentTrack(root, mp)
        node = self.getNode(root, startValue)
        path = []
        if self.solve(node, destValue, mp, path, vis):
            return ''.join(path)
        return ""
        